test/e2e/remediation_test.go

```diff
@@ -0,0 +1,477 @@
+		g.Expect(filterMachinesByPhase(machines.Items, "Running")).To(HaveLen(allMachinesCount))
+		return nil
+	}, e2eConfig.GetIntervals(specName, "wait-machine-remediation")...)
+
+	By("Scaling down machine deployment")
+	scaleMachineDeployment(ctx, bootstrapClient, 1)
+
+	By("Waiting for 2 BMHs to be Ready")
+	Eventually(func(g Gomega) error {
+		bmhs := bmh.BareMetalHostList{}
+		g.Expect(bootstrapClient.List(ctx, &bmhs, client.InNamespace(namespace))).To(Succeed())
+		g.Expect(filterBmhsByProvisioningState(bmhs.Items, bmh.StateReady)).To(HaveLen(2))
+		return nil
+	}, e2eConfig.GetIntervals(specName, "wait-machine-remediation")...).Should(Succeed())
+
+	By("Testing Metal3DataTemplate reference")
+	Logf("Creating a new Metal3DataTemplate")
+	m3dataTemplate := capm3.Metal3DataTemplate{}
+	m3dataTemplateName := fmt.Sprintf("%s-workers-template", clusterName)
+	newM3dataTemplateName := "test-new-m3dt"
+	Expect(bootstrapClient.Get(ctx, client.ObjectKey{Namespace: namespace, Name: m3dataTemplateName}, &m3dataTemplate)).To(Succeed())
+
+	newM3DataTemplate := m3dataTemplate.DeepCopy()
+	cleanObjectMeta(&newM3DataTemplate.ObjectMeta)
+
+	newM3DataTemplate.Spec.MetaData = m3dataTemplate.Spec.MetaData
+	newM3DataTemplate.Spec.NetworkData = m3dataTemplate.Spec.NetworkData
+	newM3DataTemplate.Spec.ClusterName = clusterName
+	newM3DataTemplate.Spec.TemplateReference = m3dataTemplateName
+
+	newM3DataTemplate.ObjectMeta.Name = newM3dataTemplateName
+	newM3DataTemplate.ObjectMeta.Namespace = m3dataTemplate.Namespace
+
+	err = bootstrapClient.Create(ctx, newM3DataTemplate)
+	Expect(err).NotTo(HaveOccurred())
+
+	By("Creating a new Metal3MachineTemplate")
+	m3machineTemplate := capm3.Metal3MachineTemplate{}
+	m3machineTemplateName := fmt.Sprintf("%s-workers", clusterName)
+	Expect(bootstrapClient.Get(ctx, client.ObjectKey{Namespace: namespace, Name: m3machineTemplateName}, &m3machineTemplate)).To(Succeed())
+	newM3MachineTemplateName := "test-new-m3mt"
+
+	newM3MachineTemplate := m3machineTemplate.DeepCopy()
+	cleanObjectMeta(&newM3MachineTemplate.ObjectMeta)
+
+	newM3MachineTemplate.Spec.Template.Spec.Image = m3machineTemplate.Spec.Template.Spec.Image
+	newM3MachineTemplate.Spec.Template.Spec.DataTemplate.Name = newM3dataTemplateName
+	newM3MachineTemplate.ObjectMeta.Name = newM3MachineTemplateName
+
+	Expect(bootstrapClient.Create(ctx, newM3MachineTemplate)).To(Succeed(), "Failed to create new M3MachineTemplate")
+
+	By("Pointing MachineDeployment to the new Metal3MachineTemplate")
+	deployment := clusterv1.MachineDeployment{}
+	Expect(bootstrapClient.Get(ctx, client.ObjectKey{Namespace: namespace, Name: clusterName}, &deployment)).To(Succeed())
+
+	helper, err := patch.NewHelper(&deployment, bootstrapClient)
+	Expect(err).NotTo(HaveOccurred())
+
+	deployment.Spec.Template.Spec.InfrastructureRef = v1.ObjectReference{
+		Kind:       "Metal3MachineTemplate",
+		APIVersion: "infrastructure.cluster.x-k8s.io/v1alpha5",
```

Comment:
can we parameterize the API version here?

Reply:
would be good, if possible, when we move to new API, we have to change it inside the e2e test file. But, I agree this is not only place where we need a change in e2e, so leaving it up to you to decide.